### DISCUSSION TITLE
Version 3.0.0

### DISCUSSION THREAD

**Discussion by mairas** [SignalK/SensESP#779] [Announcements]
A major new release of SensESP, including a complete rewrite of the web frontend and major restructuring of the internal class structure. Finally!

## What's Changed
* Bump version: 2.7.2 → 2.7.3-alpha by @mairas in https://github.com/SignalK/SensESP/pull/674
* Fix constant sensor configuration schema typing by @mairas in https://github.com/SignalK/SensESP/pull/684
* Remove the obsolete and broken typescript sample plugin code by @mairas in https://github.com/SignalK/SensESP/pull/685
* Add thermocouple_max6675 example by @zschramm in https://github.com/SignalK/SensESP/pull/687
* Breaking change: rename WSClient to SKWSClient by @mairas in https://github.com/SignalK/SensESP/pull/690
* Pass received Json objects always by value by @mairas in https://github.com/SignalK/SensESP/pull/691
* Improved curve interpolator by @rszemeti in https://github.com/SignalK/SensESP/pull/689
* Add a build script by @mairas in https://github.com/SignalK/SensESP/pull/695
* Patch 2 - Add NMEA2000 Output of Exhaust Gas Temperature by @zschramm in https://github.com/SignalK/SensESP/pull/693
* Compile css files automatically as part of "run build-frontend" by @mairas in https://github.com/SignalK/SensESP/pull/696
* Rename ConstantSensor getters and setters by @mairas in https://github.com/SignalK/SensESP/pull/694
* Use native ESP-IDF logging by @mairas in https://github.com/SignalK/SensESP/pull/699
* Fix empty config list by @mairas in https://github.com/SignalK/SensESP/pull/701
* Implement new ExpiringValue container class by @mairas in https://github.com/SignalK/SensESP/pull/702
* Implement new merging and repeating transforms by @mairas in https://github.com/SignalK/SensESP/pull/703
* Async Configurables by @mairas in https://github.com/SignalK/SensESP/pull/705
* Remove input_channel argument from ValueConsumer by @mairas in https://github.com/SignalK/SensESP/pull/707
* Async config handlers by @mairas in https://github.com/SignalK/SensESP/pull/710
* Make ValueConsumer::set argument a const ref by @mairas in https://github.com/SignalK/SensESP/pull/711
* Set up hostname earlier in WiFi initialization by @mairas in https://github.com/SignalK/SensESP/pull/712
* Stream producers by @mairas in https://github.com/SignalK/SensESP/pull/714
* Implement a filter transform by @mairas in https://github.com/SignalK/SensESP/pull/715
* Support more complex lists of configurables by @mairas in https://github.com/SignalK/SensESP/pull/716
* to avoid using namespace at global scope by @offspring in https://github.com/SignalK/SensESP/pull/717
* Implement a dynamic restart button by @mairas in https://github.com/SignalK/SensESP/pull/718
* Use toasts instead of modals for user feedback by @mairas in https://github.com/SignalK/SensESP/pull/719
* Use ESP-IDF native logging functions by @mairas in https://github.com/SignalK/SensESP/pull/720
* Rename web(UI) to frontend by @mairas in https://github.com/SignalK/SensESP/pull/721
* Update the README by @mairas in https://github.com/SignalK/SensESP/pull/722
* Curve interpolator fix v3 by @mairas in https://github.com/SignalK/SensESP/pull/723
* Merge dev-3 to main by @mairas in https://github.com/SignalK/SensESP/pull/724
* Release 3.0.0 beta.1 by @mairas in https://github.com/SignalK/SensESP/pull/726
* Replace mentions of Slack with Discord by @mairas in https://github.com/SignalK/SensESP/pull/732
* Fix restartRequired context handling by @mairas in https://github.com/SignalK/SensESP/pull/734
* Lint and Reformat frontend files by @mairas in https://github.com/SignalK/SensESP/pull/735
* Bump vite from 5.0.0 to 5.0.13 in /frontend-plugins/sample-plugin-js by @dependabot in https://github.com/SignalK/SensESP/pull/725
* Restore prematurely removed file by @mairas in https://github.com/SignalK/SensESP/pull/736
* Fix clang-tidy warnings by @mairas in https://github.com/SignalK/SensESP/pull/737
* Rename ReactESP classes by @mairas in https://github.com/SignalK/SensESP/pull/739
* Improve Json serialization by @mairas in https://github.com/SignalK/SensESP/pull/744
* Fix ArduinoJson deprecations by @mairas in https://github.com/SignalK/SensESP/pull/746
* Fix TimeCounter logic by @mairas in https://github.com/SignalK/SensESP/pull/747
* Invert displayMultiplier and displayOffset semantics by @mairas in https://github.com/SignalK/SensESP/pull/748
* Release 3.0.0 beta.3 by @mairas in https://github.com/SignalK/SensESP/pull/749
* Release 3.0.0 beta.3 (for reals) by @mairas in https://github.com/SignalK/SensESP/pull/750
* Fix Repeat template classes by @mairas in https://github.com/SignalK/SensESP/pull/751
* Add an event_loop() convenience function by @mairas in https://github.com/SignalK/SensESP/pull/753
* Implement a task synchronization consumable by @mairas in https://github.com/SignalK/SensESP/pull/754
* Fix event loop calls by @mairas in https://github.com/SignalK/SensESP/pull/760
* Bump vite from 5.0.13 to 5.1.8 in /frontend-plugins/sample-plugin-js by @dependabot in https://github.com/SignalK/SensESP/pull/756
* Bump rollup from 4.5.0 to 4.24.0 in /frontend-plugins/sample-plugin-js by @dependabot in https://github.com/SignalK/SensESP/pull/757
* Bump micromatch from 4.0.7 to 4.0.8 in /frontend by @dependabot in https://github.com/SignalK/SensESP/pull/758
* Bump rollup from 4.20.0 to 4.24.0 in /frontend by @dependabot in https://github.com/SignalK/SensESP/pull/759
* Web UI class cleanups by @mairas in https://github.com/SignalK/SensESP/pull/755
* Refactor web UI config item C++ implementation by @mairas in https://github.com/SignalK/SensESP/pull/761
* Rename ValueProducer::output by @mairas in https://github.com/SignalK/SensESP/pull/763
* Refactor TaskQueueProducer by @mairas in https://github.com/SignalK/SensESP/pull/764
* Unit tests and memory safety by @mairas in https://github.com/SignalK/SensESP/pull/765
* SKMetadata fix by @mairas in https://github.com/SignalK/SensESP/pull/766
* Networking compile-time configuration by @mairas in https://github.com/SignalK/SensESP/pull/767
* More smart pointers by @mairas in https://github.com/SignalK/SensESP/pull/769
* Release 3.0.0 beta.4 by @mairas in https://github.com/SignalK/SensESP/pull/770
* Improve instrumentation by @mairas in https://github.com/SignalK/SensESP/pull/772
* Release 3.0.0 beta.5 by @mairas in https://github.com/SignalK/SensESP/pull/773
* Provide correct ConfigSchema for template classes by @mairas in https://github.com/SignalK/SensESP/pull/774
* Release 3.0.0 beta.6 by @mairas in https://github.com/SignalK/SensESP/pull/775
* Introduce v3 concepts in the documentation by @mairas in https://github.com/SignalK/SensESP/pull/777
* Release v3.0.0 by @mairas in https://github.com/SignalK/SensESP/pull/778

## New Contributors
* @zschramm made their first contribution in https://github.com/SignalK/SensESP/pull/687
* @rszemeti made their first contribution in https://github.com/SignalK/SensESP/pull/689
* @offspring made their first contribution in https://github.com/SignalK/SensESP/pull/717
* @dependabot made their first contribution in https://github.com/SignalK/SensESP/pull/725

**Full Changelog**: https://github.com/SignalK/SensESP/compare/v2.7.2...v3.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/SignalK/SensESP/releases/tag/v3.0.0'>Version 3.0.0</a>.</em>

**Comment by rszemeti** [comment]
Wow ... OK, thats a fairly MAJOR release!

I have the boat out of the water and the OpenPlotter, and your bits of kit
and the engine temp sender and guage on the bench today, so this could not
have arrived at a better time!

That's sorted out what I'll be doing next then! ;)

**Reply by mairas** [reply]
Yeah, it started with reimplementation of WiFiManager, which wasn't too bad. Then, the whole frontend re-write, which made me learn React. And then it snowballed to other things -- there were a lot of things I thought could be made better but would break the API, and I thought it's better to make a sweeping change rather than a thousand small paper-cuts. I'm really happy with the result. The overall structure is cleaner and easier to work with.

I'm sure there are a lot of bugs still...
